File: problems/0041-First-Missing-Positive/sol2.cpp
// ==========================================================
// 41. First Missing Positive
// Difficulty : Hard
// Language   : C++
// Solution   : #2
// Runtime    : 111 ms (Beats 5%)
// Memory     : 84.5 MB (Beats 5%)
// Link       : https://leetcode.com/problems/first-missing-positive/
// ==========================================================

        x.insert(x.end(), y.begin(), y.end());
        if(x.size()==0) return 1;
        if(x.size()<2 && x[0]!=1) return 1; 
        int len=x[x.size()-1];
        for(int i=1;i<len;i++){
            if(i!=x[i-1]){
                return i;
            }
        }
        return x[x.size()-1]+1;
    }
};

File: problems/0042-Trapping-Rain-Water/sol2.java
// ==========================================================
// 42. Trapping Rain Water
// Difficulty : Hard
// Language   : Java
// Solution   : #2
// Runtime    : 435 ms (Beats 4%)
// Memory     : 47.8 MB (Beats 43%)
// Link       : https://leetcode.com/problems/trapping-rain-water/
// ==========================================================

class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int total = 0;
        int high1 = 0;

        while (high1 < n - 1) {
            int high2 = -1;
            for (int i = high1 + 1; i < n; i++) {
                if (height[i] >= height[high1]) {
                    high2 = i;
                    break;
                }
            }
            if (high2 == -1) {
                int maxIdx = high1 + 1;
                for (int i = high1 + 2; i < n; i++) {
                    if (height[i] > height[maxIdx]) maxIdx = i;
                }
                high2 = maxIdx;
            }

            total += count(high1, high2, height);
            high1 = high2;
        }
        return total;
    }

    int count(int high1, int high2, int[] nums) {
        int window = high2 - high1 - 1;
        if (window <= 0) return 0;
        int tempcount = window * Math.min(nums[high1], nums[high2]);
        int cc = 0;
        for (int i = high1 + 1; i < high2; i++) {  
            cc += nums[i];
        }
        return tempcount - cc;
    }
}